### PR TITLE
Address the performance issues for bulk suppression in IDE.

### LINKED ISSUE #6066: Performance issues in bulk suppressing diagnsotics from error list
Currently, it takes quite a while (>10 mins) to bulk suppress 1000+ issues from the error list. We need to improve the perf for RTM.

### PR DESCRIPTION
**Pull request by mavasani**
1. "Suppress active issues" command from solution explorer was doing couple of builds - one command line build to compute the FxCop diagnostics and another intellisense build to fetch Roslyn diagnostics. This change removes the second build and instead uses the diagnostics emitted from the build for suppression.

2. Bulk suppressing multiple diagnostics from error list/suppress active issues command was spending more than 50% of time in formatting the generated global suppressions file. This delay grows significantly larger with 1000+ issues being suppressed. This fix changes us to not format the entire suppressions file at once, but instead format each attribute separately and avoid formatting the batched file. This significantly brings down the formatting time.

3. Thread in the cancellation token during batch fix computation - otherwise cancelling the wait dialog while computing suppressions fix will make VS unresponsive.

Fixes #6066

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet-bot retest lin please

**Comment by mavasani** [comment]
@dotnet-bot retest linux please

**Comment by mavasani** [comment]
@dotnet-bot retest eta please

**Comment by mavasani** [comment]
@heejaechang @srivatsn @basoundr can you please review?

/cc @morrisjoe @michaelcfanning

**Comment by MattGertz** [comment]
Approved pending reviews.

**Comment by mavasani** [comment]
@heejaechang @srivatsn @natidea can you please take a look?

File: src/Features/CSharp/Portable/CodeFixes/Suppression/CSharpSuppressionCodeFixProvider.cs [hunk 1]
```diff
@@ -110,8 +111,7 @@ protected override SyntaxNode AddGlobalSuppressMessageAttribute(SyntaxNode newRo
             bool needsLeadingEndOfLine)
         {
             var attributeArguments = CreateAttributeArguments(targetSymbol, diagnostic);
-            var attribute = SyntaxFactory.Attribute(SyntaxFactory.ParseName(SuppressMessageAttributeName), attributeArguments)
-                .WithAdditionalAnnotations(Simplifier.Annotation);
+            var attribute = SyntaxFactory.Attribute(SyntaxFactory.ParseName(SuppressMessageAttributeName), attributeArguments);
```
Comment (heejaechang) [inline src/Features/CSharp/Portable/CodeFixes/Suppression/CSharpSuppressionCodeFixProvider.cs]: how about doing minimal type name here manually as well like you did for formatting?

File: src/Workspaces/Core/Portable/CodeActions/CodeAction.cs [hunk 2]
```diff
@@ -138,12 +138,12 @@ protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cance
...
         {
             var solution = await GetChangedSolutionAsync(cancellationToken).ConfigureAwait(false);
-            if (solution == null)
+            if (solution == null || !postProcessChanges)
             {
-                return null;
+                return solution;
             }
 
             return await this.PostProcessChangesAsync(solution, cancellationToken).ConfigureAwait(false);
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/CodeActions/CodeAction.cs]: isn't postprocesschangesasync is overridable? why have flag + overide-able method?

File: src/VisualStudio/Core/Def/Implementation/TaskList/ProjectExternalErrorReporter.cs [hunk 3]
```diff
@@ -224,10 +227,16 @@ private static DiagnosticSeverity GetDiagnosticSeverity(ExternalError error)
             string mappedFilePath, int mappedStartLine, int mappedStartColumn, int mappedEndLine, int mappedEndColumn,
             string originalFilePath, int originalStartLine, int originalStartColumn, int originalEndLine, int originalEndColumn)
         {
+            var document = id != null ? _workspace.CurrentSolution.GetDocument(id) : null;
+            var sourceSpan = document != null ?
+                document.GetSpanFromPositionAsync(originalStartLine, originalStartColumn, originalEndLine, originalEndColumn, CancellationToken.None).WaitAndGetResult(CancellationToken.None) :
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ProjectExternalErrorReporter.cs]: you shouldn't do this. by the way, are we supporting code action (including suppression) for build error? there is no way to know the build error is valid or not unless  you build before running LB like FxCop.

...

regardless of that, source span shouldn't be calculated here. getting source span is expensive since it require to bring in text.

it is done this way by design. you should call this
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Diagnostics/DiagnosticData.cs,305508f892f5f654,references

if you want TextSpan from diagnostic from build, and you should do that only when it is needed.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioDiagnosticListTable.BuildTableDataSource.cs [hunk 4]
```diff
@@ -229,6 +229,9 @@ public override bool TryGetValue(int index, string columnName, out object conten
                         case ProjectGuids:
                             content = item.ProjectGuids;
                             return ((Guid[])content).Length > 0;
+                        case SuppressionStateColumnDefinition.ColumnName:
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioDiagnosticListTable.BuildTableDataSource.cs]: how build diagnostic know whether it is suppressed or not? is out of proc build now log suppressed errors as well? I dont see diagnostic created from out of proc build ever set IsSuppressed.

are you trying to always say "Active" for diagnostic from build?

File: src/VisualStudio/Core/Def/Implementation/Extensions/DocumentExtensions.cs [hunk 5]
```diff
@@ -59,5 +60,15 @@ internal static class DocumentExtensions
 
             return codeBlocks;
         }
+
+        public async static Task<TextSpan> GetSpanFromPositionAsync(this Document document, int startLine, int startColumn, int endLine, int endColumn, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Extensions/DocumentExtensions.cs]: this doesn't work. there is no guarantee data you got from out of proc is actually valid with current solution. currently there is no way to track version of snapshot out of proc build actually used.

Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Extensions/DocumentExtensions.cs]: DiagnosticData already has logic to deal with this case with bunch of error case handling (such as span being out of index and etc)

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioSuppressionFixService.cs [hunk 6]
```diff
@@ -122,16 +123,19 @@ public bool RemoveSuppressions(bool selectedErrorListEntriesOnly, IVsHierarchy p
             }
         }
 
-        private async Task<ImmutableArray<DiagnosticData>> GetAllDiagnosticsAsync(Func<Project, bool> shouldFixInProject, CancellationToken cancellationToken)
+        private ImmutableArray<DiagnosticData> GetAllBuildDiagnostics(Func<Project, bool> shouldFixInProject, CancellationToken cancellationToken)
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioSuppressionFixService.cs]: I am not sure how this will ever work?

**Comment by heejaechang** [comment]
I dont think #1 will work unless we block vs and do explicit build + suppression and then unblock vs. and make sure the suppression from build error in the error list is disabled.

File: src/Features/CSharp/Portable/CodeFixes/Suppression/CSharpSuppressionCodeFixProvider.cs (hunk 1 above)
Comment (mavasani) [inline src/Features/CSharp/Portable/CodeFixes/Suppression/CSharpSuppressionCodeFixProvider.cs]: What API can I use to get the minimal type name? `SuppressMessageAttributeName` = `System.Diagnostics.CodeAnalysis.SuppressMessage`.

File: src/Workspaces/Core/Portable/CodeActions/CodeAction.cs (hunk 2 above)
Comment (mavasani) [inline src/Workspaces/Core/Portable/CodeActions/CodeAction.cs]: Yes, but the code action itself might override that method. This API with flag basically guarantees that the caller will get the changed solution without running Formatter or Simplifier.

File: src/VisualStudio/Core/Def/Implementation/TaskList/ProjectExternalErrorReporter.cs (hunk 3 above)
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/TaskList/ProjectExternalErrorReporter.cs]: /cc @srivatsn 

`are we supporting code action (including suppression) for build error? there is no way to know the build error is valid or not unless you build before running LB like FxCop` - Ideally, we would like to but as you mention we won't have any way to detect if it is stale or not. Also suppressing a build diagnostic from error list won't update its suppression state as we won't recompute it - so probably we should just show a blank suppression state for build diagnostics?

Thanks for pointing me to the right method to get text span - I will revert the change in this file, and compute the span in the suppression fix service when we are required to fix build diagnostics.

File: src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioDiagnosticListTable.BuildTableDataSource.cs (hunk 4 above)
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioDiagnosticListTable.BuildTableDataSource.cs]: The intention here was for build diagnostics to support suppression. Based on your other comment, I agree we should just not show any value for suppression state for build diagnostics and retain the current behavior. Let me revert this change.

File: src/VisualStudio/Core/Def/Implementation/Extensions/DocumentExtensions.cs (hunk 5 above)
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/Extensions/DocumentExtensions.cs]: Thanks, I will remove this method and use the existing method on DiagnosticData.

**Comment by mavasani** [comment]
_"I dont think #1 will work unless we block vs and do explicit build + suppression and then unblock vs. and make sure the suppression from build error in the error list is disabled."_

We already block the user on doing any operations while suppression is being computed and applied - we do so by bringing up a cancellable wait dialog. The user still has the option to modify sources while the build is going on - however, this only means the explicit build snapshot against which suppression was executed matches the snapshot on which user explicitly invoked this command, not the latest live snapshot. This is by design and no different from user waiting till build + suppression command finished and then modifying sources - you are likely to have the suppressions get out of date and stale as you change your top level definitions, and the diagnostics on the original method symbol will start showing up as active again and you need to generate new suppressions.

My point is the command "Suppress active issues" guarantees generating suppressions on the solution snapshot when the command was invoked, but not the live - and that is by design.

**Comment by mavasani** [comment]
I agree with "_make sure the suppression from build error in the error list is disabled._"

@srivatsn - do you agree? We will use build diagnostics for suppression only during explicit "Suppress active issues" command from solution explorer, but not support suppression state/error list suppression for build diagnostics.

**Comment by srivatsn** [comment]
Yes I agree.

**Comment by mavasani** [comment]
@heejaechang I have updated the PR, can you please take a look?

File: src/Features/CSharp/Portable/CodeFixes/Suppression/CSharpSuppressionCodeFixProvider.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/CSharp/Portable/CodeFixes/Suppression/CSharpSuppressionCodeFixProvider.cs]: get symbol for suppress attribute (using GetSymbolBymetadata ...) and symbol has ToMinimallyQualifed ... method in it.

also, for global suppression, since you are putting all those attribute to same context, get minimallyQualifed string once, and reuse them.

**Comment by heejaechang** [comment]
:+1:

**Comment by mavasani** [comment]
@dotnet-bot test prtest/win/dbg/unit32

**Comment by mavasani** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
